src/broadcast/scrobblingmanager.cpp

```diff
@@ -0,0 +1,215 @@
+#include <QObject>
+
+#include "broadcast/scrobblingmanager.h"
+#include "control/controlproxy.h"
+#include "engine/enginexfader.h"
+#include "mixer/deck.h"
+
+ScrobblingManager::ScrobblingManager() :
+m_CPGuiTick("[Master]", "guiTick50ms",this),
+m_CPCrossfader("[Master]","crossfader", this),
+m_CPXFaderCurve(ConfigKey(EngineXfader::kXfaderConfigKey, 
+                "xFaderCurve"),this),
+
+m_CPXFaderCalibration(ConfigKey(EngineXfader::kXfaderConfigKey, 
+                      "xFaderCalibration"),this),
+
+m_CPXFaderMode(ConfigKey(EngineXfader::kXfaderConfigKey, 
+               "xFaderMode"),this),
+
+m_CPXFaderReverse(ConfigKey(EngineXfader::kXfaderConfigKey, 
+                  "xFaderReverse"),this)
+{
+    m_CPGuiTick.connectValueChanged(SLOT(slotGuiTick(double)));
+    startTimer(1000);
+}
+
+
+void ScrobblingManager::slotTrackPaused(TrackPointer pPausedTrack) {
+    //Extra functional decision to only track main decks.            
+    Deck *sourceDeck = qobject_cast<Deck*>(sender());    
+    if (sourceDeck == 0) {
+        qDebug() << "Didn't load track in a deck yet scrobbling "
+                    "received paused signal.";
+        return;
+    }
+    // QMutexLocker locker(&m_mutex);
+    bool allPaused = true;
+    TrackInfo *pausedTrackInfo;
+    for (TrackInfo *trackInfo : m_trackList) {
+        if (trackInfo->m_pTrack == pPausedTrack) {
+            pausedTrackInfo = trackInfo;
+            for (BaseTrackPlayer *player : trackInfo->m_players) {
+                BaseTrackPlayerImpl *playerImpl = 
+                    qobject_cast<BaseTrackPlayerImpl*>(player);
+                if (playerImpl == 0) {
+                    qDebug() << "Track player interface isn't a "
+                                "BaseTrackPlayerImpl";
+                    return;
+                }
+                if (!playerImpl->isTrackPaused())
+                    allPaused = false;
+            }
+            break;
+        }                   
+    }
+    if (allPaused)
+        pausedTrackInfo->m_trackInfo.pausePlayedTime();        
+}
+
+void ScrobblingManager::slotTrackResumed(TrackPointer pResumedTrack) {
+    //Extra functional decision to only track main decks.
+    Deck *sourceDeck = qobject_cast<Deck*>(sender());    
+    if (sourceDeck == 0) {
```

Comment:
if (!sourceDeck) { or == nullptr

Reply:
I saw the specification of qobject_cast and it returns 0, I read the standard and nullptr doesn't have to be 0, same with NULL that is implementation defined. So I went with 0 because that's what the Qt doc says. Though I agree it's more readable.

Reply:
Oh, ok, then I will use !